// ang2-phaser/ng2-phaser/phaser.directive.ts
// core
import {Component, Input, Output, EventEmitter, ElementRef} from '@angular/core';


declare var Phaser:any;
//------------------------------------
@Component({
  selector: 'phaser',
   template:'',
})
export class ng2PhaserComponent {

  //--------------
  public selfRef:any;

  @Output() phaser = new EventEmitter();
  @Input() settings:any;
  

   //--------------
   constructor(private el: ElementRef) {
      this.selfRef = el.nativeElement;
   }
   //--------------

  //--------------
  ngOnInit(){
    var t = this;
    var alreadyLoaded = false;
    var allScripts = document.getElementsByTagName("script");
  
    if(t.settings == undefined){
      t.settings = {
        file: 'js/phaser.min.js'
      }
    }
 
    var js = document.createElement("script");
        js.type = "text/javascript";
        js.src = t.settings.file;
        document.body.appendChild(js);
        js.onload = function(){
            t.phaser.emit({firstLoad: true, container: t.selfRef})
        }

  }
  //--------------

  //--------------
  initPhaser(){
    this.phaser.emit({container: this.selfRef})
  }
  //--------------


}
//------------------------------------
